src/issue.cpp

```diff
@@ -171,70 +236,299 @@ Issue::ReferenceRule Issue::referenceRule() const
-    mPimpl->mCause = Issue::Cause::MODEL;
+    if (model != nullptr) {
+        setItem(ItemType::MODEL, model);
+    } else if (mPimpl->mItemType == ItemType::MODEL) {
+        mPimpl->clearItem();
+    }
 }
 
 ModelPtr Issue::model() const
 {
-    return mPimpl->mModel;
+    if (mPimpl->mItemType != ItemType::MODEL) {
+        return nullptr;
+    }
+    return std::any_cast<ModelPtr>(mPimpl->mItem);
+}
+
+void Issue::setEncapsulation(const ModelPtr &model)
+{
+    if (model != nullptr) {
+        setItem(ItemType::ENCAPSULATION, model);
+    } else if (mPimpl->mItemType == ItemType::ENCAPSULATION) {
+        mPimpl->clearItem();
+    }
+}
+
+ModelPtr Issue::encapsulation() const
+{
+    if (mPimpl->mItemType != ItemType::ENCAPSULATION) {
+        return nullptr;
+    }
+    return std::any_cast<ModelPtr>(mPimpl->mItem);
 }
 
 void Issue::setUnits(const UnitsPtr &units)
 {
-    mPimpl->mUnits = units;
-    mPimpl->mCause = Issue::Cause::UNITS;
+    if (units != nullptr) {
+        setItem(ItemType::UNITS, units);
+    } else if (mPimpl->mItemType == ItemType::UNITS) {
+        mPimpl->clearItem();
+    }
 }
 
 UnitsPtr Issue::units() const
 {
-    return mPimpl->mUnits;
+    if (mPimpl->mItemType != ItemType::UNITS) {
+        return nullptr;
+    }
+    return std::any_cast<UnitsPtr>(mPimpl->mItem);
+}
+
+void Issue::setUnit(const UnitItem &unit)
+{
+    if (unit.first != nullptr) {
+        setItem(ItemType::UNIT, unit);
+    } else if (mPimpl->mItemType == ItemType::UNIT) {
+        mPimpl->clearItem();
+    }
+}
+
+UnitItem Issue::unit() const
+{
+    if (mPimpl->mItemType != ItemType::UNIT) {
+        return std::make_pair(nullptr, 0);
+    }
+    return std::any_cast<UnitItem>(mPimpl->mItem);
+}
+
+void Issue::setConnection(const VariablePair &pair)
+{
+    if ((pair.first != nullptr) && (pair.second != nullptr)) {
+        setItem(ItemType::CONNECTION, pair);
+    } else if (mPimpl->mItemType == ItemType::CONNECTION) {
+        mPimpl->clearItem();
+    }
+}
+
+VariablePair Issue::connection() const
+{
+    if (mPimpl->mItemType != ItemType::CONNECTION) {
+        return std::make_pair(nullptr, nullptr);
+    }
+    return std::any_cast<VariablePair>(mPimpl->mItem);
+}
+
+void Issue::setMapVariables(const VariablePair &pair)
+{
+    if ((pair.first != nullptr) && (pair.second != nullptr)) {
+        setItem(ItemType::MAP_VARIABLES, pair);
+    } else if (mPimpl->mItemType == ItemType::MAP_VARIABLES) {
+        mPimpl->clearItem();
+    }
+}
+
+VariablePair Issue::mapVariables() const
+{
+    if (mPimpl->mItemType != ItemType::MAP_VARIABLES) {
+        return std::make_pair(nullptr, nullptr);
+    }
+    return std::any_cast<VariablePair>(mPimpl->mItem);
 }
 
 void Issue::setVariable(const VariablePtr &variable)
 {
-    mPimpl->mVariable = variable;
-    mPimpl->mCause = Issue::Cause::VARIABLE;
+    if (variable != nullptr) {
+        setItem(ItemType::VARIABLE, variable);
+    } else if (mPimpl->mItemType == ItemType::VARIABLE) {
```

Comment:
I'm not understanding why all these `set*()` methods will only clear the current item if the given object is a `nullptr` and the current type matches. I'm not seeing why any call to `set*()` with a `nullptr` would not cause the current item to be cleared...